In [2]:
import numpy as np
import pandas as pd
from sklearn import linear_model
import matplotlib.pyplot as plt
import statsmodels.api as sm
import warnings
warnings.filterwarnings('ignore')
house_prices_df = pd.read_csv("house_prices_train.csv")
house_prices_df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [3]:
new_features = house_prices_df[["YearBuilt","LotArea","TotRmsAbvGrd","FullBath","MoSold","HalfBath","GarageCars","SalePrice"]]
Y = new_features['SalePrice']
X = new_features[['YearBuilt','LotArea','MoSold','TotRmsAbvGrd','FullBath','HalfBath','GarageCars']]
lrm = linear_model.LinearRegression()
lrm.fit(X, Y)


X = sm.add_constant(X)
results = sm.OLS(Y, X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.604
Model:                            OLS   Adj. R-squared:                  0.602
Method:                 Least Squares   F-statistic:                     316.0
Date:                Fri, 03 Dec 2021   Prob (F-statistic):          1.83e-286
Time:                        21:39:14   Log-Likelihood:                -17868.
No. Observations:                1460   AIC:                         3.575e+04
Df Residuals:                    1452   BIC:                         3.580e+04
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const        -1.428e+06   1.15e+05    -12.454      0.000   -1.65e+06    -1.2e+06
YearBuilt      718.3683     58.871     12.202      0.000     602.887     833.849
LotArea          1.1279      0.135      8.352      0.000       0.863       1.393
MoSold         412.4068    486.515      0.848      0.397    -541.940    1366.754
TotRmsAbvGrd  1.488e+04   1119.773     13.293      0.000    1.27e+04    1.71e+04
FullBath      1.383e+04   3409.167      4.057      0.000    7143.147    2.05e+04
HalfBath      4828.7097   2924.668      1.651      0.099    -908.316    1.06e+04
GarageCars    3.284e+04   2269.839     14.468      0.000    2.84e+04    3.73e+04
==============================================================================
Omnibus:                      649.002   Durbin-Watson:                   1.984
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             7481.432
Skew:                           1.755   Prob(JB):                         0.00
Kurtosis:                      13.519   Cond. No.                     1.27e+06
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.27e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**Ev fiyatları modelinizi tekrar çalıştırın ve F-testi, R-kare, ayarlanmış R-kare, AIC ve BIC kullanarak modelinizin uygunluğunu değerlendirin.**

F-testi, R-kare, ayarlanmış R-kare, AIC ve BIC değerlerine baktığımızda F-Testi bir miktar düşük çıktı. Önceki ödevden elde ettiğim kanıya göre alakasız feature'ları çıkardığımda bu değerin yükseldiğini gördüm.

R-Kare değeri %70 in üzerine çıktığında model bizim için iyi diyebiliriz. Ancak burada %60 da kaldı. Orta derecede kaliteli bir modelimiz var diyebiliriz. Yani modelimizde bağımsız değişken bağımlı değişken hakkında %60 oranında tahminde bulunabiliyor.

Ayarlanmış R-kare değerinin R-kare ile birbirlerine çok yakın olduğunu görüyoruz. Sadece R-kare değeri 0.604 iken o 0.602 de kalmış. ayarlanmış R-kare değeri bizi korele değer varsa cezalandırır ve yavaş yükselir. Buradan da veri setimizde ufak bir korelasyon vardır diyebiliriz. Ancak bu değer çok da yüksek değildir çünkü  R-kare değerine hayli yakın çıkmış.

AIC ve BIC değerleri ne kadar düşük olursa o kadar iyidir bizim için. Muhtemelen buradaki p value'ları 0.05 den büyük olan değişkenleri veri setinden çıkarırsak bu değerler daha da çok düşecektir.

**Modelinizin tatmin edici olduğunu düşünüyor musunuz? Açıklayın**

Modelin tatmin edici olduğunu düşünmüyorum. 80 civarı feature varken, mutlaka aralarında hedef değişkenimizi daha iyi tahmin etmemize yarayacak bir değişken bulabiliriz. Bu R-kare değeri daha da yükselebilir. Yüzde 70in üzerinde çıkmaya başladığında bizim için iyidir diyebiliriz.

**Modelinizin uygunluğunu iyileştirmek için, bazı değişkenleri ekleyerek veya kaldırarak farklı model özelliklerini deneyin.**

OverallQual: Rates the overall material and finish of the house (Evin genel malzemesini ve bitişini değerlendirmesi)

OverallCond: Rates the overall condition of the house (Evin genel durumunun değerlendirilmesi)

1stFlrSF: First Floor square feet  (1. katın fit karesi)
 
2ndFlrSF: Second floor square feet (2. katın fit karesi)

GarageArea: Size of garage in square feet (Garajın fit karesi)


In [4]:
new_features = house_prices_df[['YearBuilt','LotArea','OverallQual','TotRmsAbvGrd','FullBath','OverallCond','GarageCars','1stFlrSF','2ndFlrSF','GarageArea',"SalePrice"]]
Y = new_features['SalePrice']
X = new_features[['YearBuilt','LotArea','OverallQual','TotRmsAbvGrd','FullBath','OverallCond','GarageCars','1stFlrSF','2ndFlrSF','GarageArea']]
lrm = linear_model.LinearRegression()
lrm.fit(X, Y)


X = sm.add_constant(X)
results = sm.OLS(Y, X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.780
Model:                            OLS   Adj. R-squared:                  0.778
Method:                 Least Squares   F-statistic:                     512.3
Date:                Fri, 03 Dec 2021   Prob (F-statistic):               0.00
Time:                        21:57:04   Log-Likelihood:                -17440.
No. Observations:                1460   AIC:                         3.490e+04
Df Residuals:                    1449   BIC:                         3.496e+04
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const        -1.073e+06   9.62e+04    -11.156      0.000   -1.26e+06   -8.85e+05
YearBuilt      487.1267     49.237      9.894      0.000     390.544     583.710
LotArea          0.6324      0.105      6.041      0.000       0.427       0.838
OverallQual   2.124e+04   1139.754     18.633      0.000     1.9e+04    2.35e+04
TotRmsAbvGrd   159.9387   1091.047      0.147      0.883   -1980.263    2300.140
FullBath     -5780.1662   2618.203     -2.208      0.027   -1.09e+04    -644.292
OverallCond   6269.1782    975.541      6.426      0.000    4355.555    8182.802
GarageCars    9604.7644   2996.357      3.205      0.001    3727.102    1.55e+04
1stFlrSF        77.2705      4.626     16.704      0.000      68.196      86.345
2ndFlrSF        49.2352      4.220     11.668      0.000      40.958      57.513
GarageArea      14.7686     10.149      1.455      0.146      -5.140      34.677
==============================================================================
Omnibus:                      442.513   Durbin-Watson:                   1.997
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            54541.963
Skew:                          -0.193   Prob(JB):                         0.00
Kurtosis:                      32.940   Cond. No.                     1.43e+06
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.43e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**Denediğiniz her model için, uygun metrikleri alın ve modellerinizi birbiriyle karşılaştırın. Hangi model en iyisidir ve neden?**

İlk modelin metrikleri ;

**R-squared:**	0.604 __ | __   **Adj. R-squared:**	0.602  __ | __ **F-statistic:**	316.0  __ |__  **AIC**:3.575e+04 ___ | __   **BIC**:3.580e+04



İkinci modelin metrikleri;

**R-squared:**	0.780 __ | __   **Adj. R-squared:**	0.778  __ | __ **F-statistic:**	512.3  __ |__  **AIC**:3.490e+04 ___ | __   **BIC**:3.496e+04 
 

İki modeli de karşılaştırdığımızda görülüyor ki MoSold ve HalfBath özelliklerini çıkarıp yerlerine OverallQual, 1stFlrSF, 2ndFlrSF, GarageArea, OverallCond özelliklerini koyduğumuzda modelimizin R-kare değeri bayağı arttı. %60 dan %78'e çıktı. Bu güzel bir artış. Ancak p-value değerlerine baktığımızda TotRmsAbvGrd özelliğinin bu değerinin arttığını görüyoruz. Demek ki veri setinde bu feature'a korele değer/değerler var. Aynı zamanda GarageArea feature'unun da p-value değeri 0.05'in üzerinde çıkmış. Bizim için çok anlamsız olmasa da bir miktar 'hedef değişkeni açıklama'da yetersiz. 
Adj. R-kare değerine baktığımızda R-kare değerinin bir miktar gerisinde kaldığını görüyoruz. Bu da bize korele değerler olduğu için modeli cezalandırdığını söylemekte. F-statistic değeri yükselmiş. Bu da istediğimiz gibi olmuş. AIC ve BIC değerlerinin de düştüğü gözlemlenmekte. Bu da modelimizin iyileştiğini göstermektedir.

Bu sebeplerle 2. model daha iyidir.